src/org/rascalmpl/library/analysis/diff/edits/HiFiLayoutDiff.rsc

```diff
@@ -0,0 +1,120 @@
+    assert original := formatted : "nothing except layout and keyword fields may be different for layoutDiff to work correctly.";
+
+    @synopsis{rec is the recursive workhorse, doing a pairwise recursion over the original and the formatted tree}
+    @description{
+        We recursively skip over every "equal" pairs of nodes, until we detect two different _layout_ nodes. The original location
+        of that node and the new contents of the formatted node is used to construct a replace ((TextEdit)), and 
+        optionally the original layout is inspected for source code comments which may have been lost. Literals are skipped 
+        explicitly to avoid arbitrary edits for case insensitive literals, and to safe some time.
+    }
+
+    // if layout differences are detected, here we produce a `replace` node:
+    list[TextEdit] rec(
+        t:appl(prod(Symbol tS, _, _), list[Tree] tArgs), // layout is not necessarily parsed with the same rules (i.e. comments are lost!)
+        u:appl(prod(Symbol uS, _, _), list[Tree] uArgs))
+        = [replace(t@\loc, copyComments ? learnComments(t, "<u>") : "<u>") | tArgs != uArgs] 
+        when 
+            delabel(tS) is layouts, 
+            delabel(uS) is layouts,
+            tArgs != uArgs; 
+
+    // matched literal trees generate empty diffs
+    list[TextEdit] rec(
+        appl(prod(lit(_), _, _), list[Tree] _),
+        appl(prod(lit(_), _, _), list[Tree] _))
+        = [];
+
+    // matched case-insensitive literal trees generate empty diffs such that the original is maintained
+    list[TextEdit] rec(
+        appl(prod(cilit(_), _, _), list[Tree] _),
+        appl(prod(cilit(_), _, _), list[Tree] _))
+        = [];
+
+    list[TextEdit] rec(
+        char(_),
+        char(_)
+    ) = [];
+
+    list[TextEdit] rec(
+        cycle(Symbol _, int _),
+        cycle(_, _)
+    ) = [];
+
+    // recurse through the entire parse tree to collect layout edits:
+    default list[TextEdit] rec(
+        Tree t:appl(Production p, list[Tree] argsA),
+        appl(p /* must be the same by the above assert */, list[Tree] argsB)) 
+        = [*rec(a, b) | <a, b> <- zip2(argsA, argsB)]; 
+
+    
+
+    // first add required locations to layout nodes
+    original = reposition(original, markLit=true, markLayout=true, markSubLayout=true);
+
+    return rec(original, formatted);
+}
+
+    
+@synopsis{Make sure the new layout still contains all the source code comments of the original layout}
+@description{
+This algorithm uses the @category("Comments") tag to detect source code comments inside layout substrings. If the original
+layout contains comments, we re-introduce the comments at the expected level of indentation. New comments present in the 
+replacement are also kept.
+
+This trick is complicated by the syntax of multiline comments and single line comments that have
+to end with a newline.
+}
+private str learnComments(Tree original, str replacement) {
```

Comment:
This code duplicates source code comments if they occur both in the original and the replacement.